### PR TITLE
fix #8743 by using interface instead of Network object

### LINKED ISSUE #8743: panic: interface conversion: object.NetworkReference is *object.DistributedVirtualPortgroup, not *object.Network
#### Overview of the Issue

Creating vsphere-clone of existing template made by packer vsphere-iso previously.  App panics immediately on step 1: Cloning VM ... with unexpected EOF.

#### Reproduction Steps

packer validate XXX 
packer build XXX

### Packer version

Packer v1.5.2

### Simplified Packer Buildfile

{
  "builders": [
    {
      "type": "vsphere-clone",
      "cluster": "{{ user `cluster` }}",
      "convert_to_template": "true",
      "datastore": "{{ user `datastore` }}",
      "insecure_connection": "true",
      "network": "{{ user `vsphere_network` }}",
      "username": "{{ user `vsphere_user` }}",
      "password": "{{ user `vsphere_password` }}",
      "ram": "1024",
      "resource_pool": "{{ user `resource_pool` }}",
      "vcenter_server": "{{ user `vsphere_server` }}",
      "vm_name": "{{ user `build_name` }}",
      "communicator": "ssh",
      "cpus": "1",
      "disk_size": "30720",
      "pause_before_connecting": "2m",
      "shutdown_command": "echo 'packer' | sudo -S shutdown -P now",
      "ssh_password": "{{ user `packer_password` }}",
      "ssh_username": "packer",
      "template": "{{ user `vmware_template` }}"
    }
  ],
yada, yada, yada

### Operating system and Environment details

VMWare RHEL 7.6

### Log Fragments and crash.log files

Template validated successfully.
vsphere-clone: output will be in this color.

==> vsphere-clone: Cloning VM...
Build 'vsphere-clone' errored: unexpected EOF

==> Some builds didn't complete successfully and had errors:
--> vsphere-clone: unexpected EOF

==> Builds finished but no artifacts were created.
panic: interface conversion: object.NetworkReference is *object.DistributedVirtualPortgroup, not *object.Network
2020/02/14 10:35:43 packer.1.5.2 plugin:
2020/02/14 10:35:43 packer.1.5.2 plugin: goroutine 161 [running]:
...
2020/02/14 10:35:43 [INFO] (telemetry) ending vsphere-clone
2020/02/14 10:35:43 ui error: Build 'vsphere-clone' errored: unexpected EOF
2020/02/14 10:35:43 /usr/local/bin/packer.1.5.2: plugin process exited
2020/02/14 10:35:43 machine readable: error-count []string{"1"}
2020/02/14 10:35:43 ui error:

### PR DESCRIPTION
**Pull request by jhawk28**
this should fix the panic

Closes #8743

### CODE REVIEW

File: builder/vsphere/driver/network.go [hunk 1]
```diff
@@ -37,7 +37,9 @@ func (n *Network) Info(params ...string) (*mo.Network, error) {
 		p = params
 	}
 	var info mo.Network
-	err := n.network.Properties(n.driver.ctx, n.network.Reference(), p, &info)
+
+	network := n.network.(*object.Network)
```
Comment (sylviamoss) [inline builder/vsphere/driver/network.go]: Is there any chance of a panic here?

Comment (jhawk28) [reply builder/vsphere/driver/network.go]: I wasn't 100% sure what to do here. The current implementation only calls Info as part of the builder "test". That test only calls Info after running .NewNetwork(). This should only result in the *object.Network{}. When I looked into the NetworkReference interface, there is not a generic .Properties() function even though they all have it (from Common). The data that you want ends up being passed in (&info in this case). If you have a VirtualSwitch or DVSPortGroup, you would be passing in a different managed object.

**Comment by jhawk28** [comment]
Is there a way to get the binaries from circleci? I suspect that the user would need to test it in their environment.

**Comment by sylviamoss** [comment]
Sure! One of the checks under `All checks have passed` is the store_artifacts pipeline. There you can find the binaries. Here's the link: https://circleci.com/gh/hashicorp/packer/34731#artifacts/containers/0

File: builder/vsphere/driver/network.go (hunk 1 above)
Comment (azr) [reply builder/vsphere/driver/network.go]: I'd may be just add a check just for the sake of it:
```suggestion
	network, ok := n.network.(*object.Network)
	if !ok {
		return nil, fmt.Errorf("unexpected %t network object type", n.network)
	}
```

Comment (azr) [reply builder/vsphere/driver/network.go]: One nicer thing you could do here is create a new interface:

```go
type NetworkReference interface {
	object.NetworkReference
	Properties(ctx context.Context, r types.ManagedObjectReference, ps []string, dst interface{}) error
}

type Network struct {
	driver  *Driver
	network NetworkReference
}
```

So you don't have to type-check (cast) anything

Comment (jhawk28) [reply builder/vsphere/driver/network.go]: The interface is going to just move the problem. You need to pass in a specific dst interface depending on what sort of object that you are starting with.

**Pull request merged**